feat: implement GetDependencies method for api_v3 QueryService

Harsh-6291 opened the pull request:
## Description
Implements the GetDependencies method for the Jaeger API v3 QueryService, allowing clients to query service dependency links over a specified time range.

## Changes Made
- Add new handler package in `cmd/query/app/querysvc/v3api/`
- Implement `GetDependencies` RPC method with full validation
- Add `GetDependencyReader()` getter method to v2 QueryService
- Comprehensive test coverage with 8 test cases:
  - Success cases with valid data
  - Empty result handling
  - Missing parameter validation
  - Invalid time ranges
  - Storage error handling

## Testing
✅ All tests passing:
- `TestGetDependencies_Success`
- `TestGetDependencies_EmptyResult`
- `TestGetDependencies_MissingEndTime`
- `TestGetDependencies_MissingLookback`
- `TestGetDependencies_NegativeLookback`
- `TestGetDependencies_FutureEndTime`
- `TestGetDependencies_StorageError`
- `TestGetDependencies_TimeRangeBeforeEpoch`

✅ Build verified: `go build ./cmd/query/app/...` - SUCCESS

## Related Issues
Closes #7595

## Checklist
- [x] Code follows project conventions
- [x] Tests added and passing
- [x] Proto definitions included
- [x] Request validation implemented
- [x] Error handling comprehensive
- [x] No breaking changes

Harsh-6291 committed:
feat: implement GetDependencies method for api_v3 QueryService

- Add GetDependencies gRPC handler in cmd/query/app/querysvc/v3api/
- Implement request validation for end_time and lookback parameters
- Add comprehensive test coverage with 8 test cases
- Support for proto-defined GetDependenciesRequest/Response types
- Proper error handling and logging
- Add GetDependencyReader getter to v2 QueryService
- Update server.go to register v3api handlers

Signed-off-by: harsh6921 <pachorihimanshu0@gmail.com>

yurishkuro commented on cmd/query/app/querysvc/v3api/handler.go:
```diff
@@ -0,0 +1,131 @@
+)
+
+// Handler implements api_v3.QueryServiceServer
+type Handler struct {
```
all handlers are already defined in cmd/query/app/apiv3

Harsh-6291 committed:
feat: migrate GetDependencies to new apiv3 location

- Add GetDependencies HTTP endpoint to cmd/jaeger/internal/extension/jaegerquery/internal/apiv3/http_gateway.go
- Endpoint: GET /api/v3/dependencies?end_time=<RFC3339>&lookback=<duration>
- Returns JSON with dependency links
- Note: gRPC handler not added as api_v3 proto lacks GetDependencies definitions

Signed-off-by: harsh6921 <pachorihimanshu0@gmail.com>

Harsh-6291 committed:
chore: remove obsolete v3api files from old location

- Delete cmd/query/app/apiv3/grpc_handler_dependencies_test.go
- Delete cmd/query/app/querysvc/v3api/handler.go
- Delete cmd/query/app/querysvc/v3api/handler_test.go

These files are now obsolete after migrating GetDependencies to the new
cmd/jaeger/internal/extension/jaegerquery/internal/apiv3/ location.

Signed-off-by: harsh6921 <pachorihimanshu0@gmail.com>

Harsh-6291 committed:
test: add comprehensive tests for GetDependencies HTTP endpoint

- Add TestHTTPGatewayGetDependencies for successful dependency retrieval
- Add TestHTTPGatewayGetDependenciesErrors for error cases
  - Missing end_time parameter
  - Missing lookback parameter
  - Invalid end_time format
  - Invalid lookback format
- All 5 tests passing

Signed-off-by: harsh6921 <pachorihimanshu0@gmail.com>

yurishkuro reviewed (changes requested):
comments not addressed

Harsh-6291 committed:
feat: migrate GetDependencies to new architecture and address feedback

- Migrate GetDependencies HTTP endpoint to cmd/jaeger/internal/extension/jaegerquery/internal/apiv3/
- Add comprehensive tests for GetDependencies HTTP gateway
- Remove unnecessary GetDependencyReader() from QueryService as requested by maintainers
- Address rebase requirements and move logic to the official v3 location

Signed-off-by: harsh6921 <pachorihimanshu0@gmail.com>

Harsh-6291 committed:
feat: address copilot review comments (validation, error handling, tests)

- Improve validation error messages to be more specific
- Add error handling for JSON response encoding
- Fix GetDependencies mock expectation in tests (3 args)
- Add test case for storage error handling (500 Internal Server Error)

Signed-off-by: harsh6921 <pachorihimanshu0@gmail.com>

yurishkuro commented on jaeger-ui:
ui module change does not belong here

Harsh-6291 committed:
feat: simplify validation logic

- Replace complex switch statement with single error message
- Keep validation simple: just check if params are empty and return 'both X and Y are required'
- Revert accidental jaeger-ui submodule update

Signed-off-by: harsh6921 <pachorihimanshu0@gmail.com>

Harsh-6291 committed:
Merge branch 'main' into feat/add-getdependencies-v3api

Signed-off-by: Harsh-6291 <pachorihimanshu0@gmail.com>

Harsh-6291 committed:
Merge branch 'main' into feat/add-getdependencies-v3api

Harsh-6291 committed:
Implement GetDependencies for API v3 (gRPC & HTTP)

Signed-off-by: harsh6921 <pachorihimanshu0@gmail.com>

Harsh-6291 committed:
feat: fix GetDependencies RPC implementation review feedback

- Refactor HTTP gateway to handle generic router interface and fix tracing
- Correct mock signatures and context matching in tests
- Remove redundant QueryService methods and strip UTF-8 BOM
- Add missing BasePath field to HTTPGateway struct
- Fix parameter validation and error handling in HTTP gateway

Signed-off-by: harsh6921 <pachorihimanshu0@gmail.com>

Harsh-6291 committed:
Merge branch 'main' into feat/add-getdependencies-v3api

yurishkuro committed:
Merge branch 'main' into feat/add-getdependencies-v3api

yurishkuro commented on cmd/jaeger/internal/extension/jaegerquery/internal/apiv3/http_gateway.go:
```diff
@@ -287,14 +302,63 @@ func (h *HTTPGateway) getOperations(w http.ResponseWriter, r *http.Request) {
+func (h *HTTPGateway) getDependencies(w http.ResponseWriter, r *http.Request) {
+	q := r.URL.Query()
+	endTimeStr := q.Get(paramEndTime)
+	lookbackStr := q.Get(paramLookback)
```
the IDL accepts start and end time. The Web API must match it, e.g. see `idl/swagger/api_v3/query_service.openapi.yaml`

yurishkuro replied:
rebase after #8551 is merged

Harsh-6291 committed:
refactor(jaegerquery): Address PR feedback for GetDependencies v3 API

Signed-off-by: harsh6921 <pachorihimanshu0@gmail.com>

Harsh-6291 committed:
chore: Merge upstream/main into feat/add-getdependencies-v3api

Signed-off-by: harsh6921 <pachorihimanshu0@gmail.com>

yurishkuro commented:
needs a rebase

Harsh-6291 committed:
Merge branch 'main' into feat/add-getdependencies-v3api

Harsh-6291 committed:
fix(apiv3): Remove redundant spanNameHandler from addRoute

otelhttp.NewHandler already sets the span name from the route
parameter, so wrapping with spanNameHandler on the outside was
both redundant and broken - it ran before the OTel span was
created, making trace.SpanFromContext return a no-op span.

Removes spanNameHandler call from addRoute and deletes the
unused helper function.

Signed-off-by: harsh6921 <pachorihimanshu0@gmail.com>

Harsh-6291 commented:
hey @yurishkuro  PTAL

Harsh-6291 committed:
test(apiv3): Remove unused basePath parameter from test helpers

Signed-off-by: harsh6921 <pachorihimanshu0@gmail.com>

yurishkuro reviewed (changes requested):
This PR has become unreviewable. I recommend smaller PRs (grpc and http separately) changing only what's necessary.

Harsh-6291 commented:
@yurishkuro Thanks for the feedback and guidance! I  agree that splitting this up makes it much more reviewable.

I am closing this PR to start fresh. I've just opened the first part  gRPC implementation **https://github.com/jaegertracing/jaeger/pull/8745** please take a look at this

Once the gRPC side looks good , I will follow up with the second PR for the HTTP Gateway implementation. 
Thanks again for your patience and review!
